cardano-testnet/test/Spec/Cli/Babbage/LeadershipSchedule.hs

```diff
@@ -0,0 +1,202 @@
+{-# LANGUAGE BlockArguments           #-}
```

Comment:
Why?

Reply:
I discovered it was an interesting way to use layout in expressions instead of parenthesis.

For example the expression:

```
  H.rewriteYamlFile (tempAbsPath </> "configuration.yaml") . J.rewriteObject
    $ HM.delete "GenesisFile"
    . HM.insert "Protocol" (J.toJSON @String "Cardano")
    ...
    . HM.insert "TestEnableDevelopmentNetworkProtocols" (J.toJSON True)
    . flip HM.alter "setupScribes"
      ( fmap
        ( J.rewriteArrayElements
          ( J.rewriteObject
            ( HM.insert "scFormat"
              ( case nodeLoggingFormat testnetOptions of
                  NodeLoggingFormatAsJson -> "ScJson"
                  NodeLoggingFormatAsText -> "ScText")))))
```

Can be written as:

```
  H.rewriteYamlFile (tempAbsPath </> "configuration.yaml") . J.rewriteObject
    $ HM.delete "GenesisFile"
    . HM.insert "Protocol" (J.toJSON @String "Cardano")
    ...
    . HM.insert "TestEnableDevelopmentNetworkProtocols" (J.toJSON True)
    . flip HM.alter "setupScribes" do
        fmap do
          J.rewriteArrayElements do
            J.rewriteObject do
              HM.insert "scFormat" case nodeLoggingFormat testnetOptions of
                NodeLoggingFormatAsJson -> "ScJson"
                NodeLoggingFormatAsText -> "ScText"
```

This works because `do x y z` is the same as `x y z`.  This is true even without `BlockArguments`.

What `BlockArguments` gives us is that in the expression `f $ x y z` can be rewritten as `f do x y z`.

This gives us two nice things:

* We can use `do` to use layout to group expressions, which means we don't have to match parens anymore.  When we write Haskell, we usually indent sub-expressions to help the reader read the code anyway.  The parentheses are for the compiler because layout is not significant to it inside an expression.  `BlockArguments` allows code to be written such that the mechanism humans use to group expressions and the compiler uses to group expressions can be one and the same.
* The use of `$` is unreliable because it is an operator and operators have a precedence and sometimes it is the wrong precedence.

For example if you tried to write the large expression from before using `$` instead it is actually wrong and doesn't compile:

```
  H.rewriteYamlFile (tempAbsPath </> "configuration.yaml") . J.rewriteObject
    $ HM.delete "GenesisFile"
    . HM.insert "Protocol" (J.toJSON @String "Cardano")
    ...
    . HM.insert "TestEnableDevelopmentNetworkProtocols" (J.toJSON True)
    . flip HM.alter "setupScribes"
      $ fmap
        $ J.rewriteArrayElements
          $ J.rewriteObject
            $ HM.insert "scFormat"
              $ case nodeLoggingFormat testnetOptions of
                  NodeLoggingFormatAsJson -> "ScJson"
                  NodeLoggingFormatAsText -> "ScText"
```

Similarly, `f $ case { x -> y }` can be rewritten without the `$` as `f case {x -> y}`.

And `f $ \x -> y` can be rewritten without the `$` as `f \x -> y`.

Basically there are a number of cases where when we previously used `$`, it can simply be dropped.

Reply:
You could write it as follows:
```
        (fmap
          . J.rewriteArrayElements
            . J.rewriteObject
               . HM.insert "scFormat" $ case nodeLoggingFormat testnetOptions of
                                        NodeLoggingFormatAsJson -> "ScJson"
                                        NodeLoggingFormatAsText -> "ScText")
 ```
 The problem I have with `BlockArguments` is that it makes is more difficult to discern between monadic actions and function composition. For this reason I avoid using it.

Reply:
I've taken your suggestion, as it's really only two places.

Mainly I thought that `BlockArguments` has some really interesting properties and that it is a better `$` and wanted to shared the idea.

At the place I worked previously, we used block argument `do` for error handling in monadic contexts, so perhaps we can revisit this.

Reply:
> we used block argument do for error handling in monadic contexts

Do you have an example of this I can look at?

Reply:
I do, but I'll figure out an appropriate time to share it.  It can wait until some urgent features are done.